pkg/controllers/csi/metadata/sqlite_gorm_client.go

```diff
@@ -0,0 +1,197 @@
+
+	CreateCodeModule(ctx context.Context, codeModule *CodeModule) error
+	ReadCodeModuleByVersion(ctx context.Context, version string) (*CodeModule, error)
+	DeleteCodeModule(ctx context.Context, codeModule *CodeModule) error
+
+	CreateOSMount(ctx context.Context, osMount *OSMount) error
+	UpdateOSMount(ctx context.Context, osMount *OSMount) error
+	DeleteOSMount(ctx context.Context, osMount *OSMount) error
+	ReadOSMountByTenantUUID(ctx context.Context, tenantUUID string) (*OSMount, error)
+
+	CreateAppMount(ctx context.Context, appMount *AppMount) error
+	UpdateAppMount(ctx context.Context, appMount *AppMount) error
+	DeleteAppMount(ctx context.Context, appMount *AppMount) error
+	ReadAppMountByVolumeMetaID(ctx context.Context, volumeMetaID string) (*AppMount, error)
+	ReadAppMounts(ctx context.Context) ([]AppMount, error)
+}
+
+type DBConn struct {
+	db *gorm.DB
+}
+
+var _ DBAccess = (*DBConn)(nil)
+
+// NewDBAccess creates a new gorm db connection to the database.
+func NewDBAccess(path string) (DBConn, error) {
+	// we should explicitly enable foreign_keys for sqlite
+	if strings.Contains(path, "?") {
+		path += "&_foreign_keys=on"
+	} else {
+		path += "?_foreign_keys=on"
+	}
+
+	db, err := gorm.Open(sqlite.Open(path), &gorm.Config{Logger: logger.Default})
+
+	if err != nil {
+		return DBConn{}, err
+	}
+
+	return DBConn{db: db}, nil
+}
+
+// SchemaMigration runs gormigrate migrations to create tables
+func (conn *DBConn) SchemaMigration(_ context.Context) error {
+	m := gormigrate.New(conn.db, gormigrate.DefaultOptions, []*gormigrate.Migration{})
+	m.InitSchema(func(tx *gorm.DB) error {
+		err := tx.AutoMigrate(
+			&TenantConfig{},
+			&CodeModule{},
+			&OSMount{},
+			&AppMount{},
+			&VolumeMeta{},
+		)
+		if err != nil {
+			return err
+		}
+		// all other constraints, indexes, etc...
+		return nil
+	})
+
+	_ = m.Migrate()
+
+	return gormigrate.New(conn.db, gormigrate.DefaultOptions, []*gormigrate.Migration{
+		{
+			ID:      "202403041200",
+			Migrate: dataMigration,
+			Rollback: func(tx *gorm.DB) error {
+				return nil
+			},
+		},
+	}).Migrate()
+}
+
+func (conn *DBConn) CreateTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error {
+	return conn.db.WithContext(ctx).Create(tenantConfig).Error
+}
+
+func (conn *DBConn) UpdateTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error {
+	return conn.db.WithContext(ctx).Updates(tenantConfig).Error
+}
+
+func (conn *DBConn) DeleteTenantConfig(ctx context.Context, tenantConfig *TenantConfig) error {
+	return conn.db.WithContext(ctx).Delete(tenantConfig).Error
+}
+
+func (conn *DBConn) CreateCodeModule(ctx context.Context, codeModule *CodeModule) error {
+	return conn.db.WithContext(ctx).Create(codeModule).Error
+}
+
+func (conn *DBConn) ReadCodeModuleByVersion(ctx context.Context, version string) (*CodeModule, error) {
+	var codeModule CodeModule
+
+	result := conn.db.WithContext(ctx).First(&codeModule, "version = ?", version)
```

Comment:
That won't work as stated above: in case of invalid tenantUUID it will create an invalid object instead of returning an error- at least that's what my tests showed.